shell/server/hack-session.js

```diff
@@ -43,24 +43,40 @@ SessionContextImpl = class SessionContextImpl {
     this.identityId = identityId;
   }
 
-  offer(cap, requiredPermissions) {
-    const _this = this;
+  offer(cap, requiredPermissions, descriptor, displayInfo) {
     return inMeteor(() => {
-      if (!_this.identityId) {
+      if (!this.identityId) {
         // TODO(soon): allow non logged in users?
         throw new Meteor.Error(400, 'Only logged in users can offer capabilities.');
       }
 
       const castedCap = cap.castAs(SystemPersistent);
-      const save = castedCap.save({webkey: null});
+      let apiTokenOwner = {webkey: null};
+      const isUiView = descriptor && descriptor.tags && descriptor.tags.length === 1 &&
+          descriptor.tags[0] && descriptor.tags[0].id &&
+          descriptor.tags[0].id === Grain.UiView.typeId;
+      if (isUiView) {
+        apiTokenOwner = {
+          user: {
+            identityId: this.identityId,
+            // The following fields will be overwritten by PersistentUiView.save(), so no need to
+            // pass them in:
+            //lastUsed: new Date(),
+            //title: "", // This will be replaced by the token's title
+            //denormalizedGrainMetadata: {}, // This will look up the package for the grain referenced.
+          },
+        };
+      }
+
+      const save = castedCap.save(apiTokenOwner);
```

Comment:
I realize that this isn't part of your change here, but I just noticed this gap between `save()` and the addition of the membrane requirements. Ideally they would happen as a single atomic operation, right? Actually, it looks like what happens now is if the membrane requirement fails, then the token gets persisted without any membrane requirements, because this method exists early. Am I missing something here? It seems like, at the very least, the `save()` should go directly before the `ApiTokens.update()`. cc @jparyani